core/src/impl/Kokkos_Half_MathematicalFunctions.hpp

```diff
@@ -0,0 +1,214 @@
+#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT
+#define KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(FUNC)          \
+  KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t FUNC( \
+      Kokkos::Experimental::half_t x) {                     \
+    return static_cast<Kokkos::Experimental::half_t>(       \
+        Kokkos::FUNC(static_cast<float>(x)));               \
+  }
+
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(fabs)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(exp)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(exp2)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(expm1)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(log)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(log10)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(log2)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(log1p)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(sqrt)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(cbrt)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(sin)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(cos)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(tan)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(asin)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(acos)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(atan)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(sinh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(cosh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(tanh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(asinh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(acosh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(atanh)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(erf)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(erfc)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(tgamma)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(lgamma)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(ceil)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(floor)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(trunc)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(round)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(nearbyint)
+KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF(logb)
+
+#undef KOKKOS_IMPL_MATH_UNARY_FUNCTION_HALF
+
+#define KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(FUNC)                     \
+  KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t FUNC(             \
+      Kokkos::Experimental::half_t x, Kokkos::Experimental::half_t y) { \
+    return static_cast<Kokkos::Experimental::half_t>(                   \
+        Kokkos::FUNC(static_cast<float>(x), static_cast<float>(y)));    \
+  }
+
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(fmod)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(remainder)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(fmax)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(fmin)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(fdim)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(pow)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(hypot)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(atan2)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(nextafter)
+KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF(copysign)
+
+#undef KOKKOS_IMPL_MATH_BINARY_FUNCTION_HALF
+
+KOKKOS_INLINE_FUNCTION Kokkos::Experimental::half_t abs(
+    Kokkos::Experimental::half_t x) {
+  return static_cast<Kokkos::Experimental::half_t>(
+      Kokkos::abs(static_cast<float>(x)));
+}
```

Comment:
In this case abs isn't different from the other Unary is it? Abs is a special case on the underlying math functions due to the way how abs is part of the pure integer stuff and the cmath in the standard, but it is not here a special case or?